lib/db_helper.rb

```diff
@@ -10,4 +10,25 @@ def self.execute_sql(sql)
     raw_connection.set_server_option(Mysql2::Client::OPTION_MULTI_STATEMENTS_OFF)
     raw_connection.abandon_results!
   end
+
+  def self.with_temp_table(table_name)
+    temp_table_name = "#{table_name}_temp"
+    old_table_name = "#{table_name}_old"
+
+    ActiveRecord::Base.connection.execute("CREATE TABLE #{temp_table_name} LIKE #{table_name}")
+    ActiveRecord::Base.connection.execute("ALTER TABLE #{temp_table_name} AUTO_INCREMENT = 1")
+
+    ActiveRecord::Base.transaction do
+      yield temp_table_name
+    end
+
+    # Atomically swap the tables
+    ActiveRecord::Base.connection.execute("RENAME TABLE #{table_name} TO #{old_table_name}, #{temp_table_name} TO #{table_name}")
```

Comment:
Just curious - why don't we instead delete the table_name and rename the temp_table to table_name (in a single transaction)?

Reply:
Because in most database systems, statements that fundamentally alter the structure of the DB (like `CREATE TABLE`, `DROP TABLE`, etc.) implicitly cause a transaction commit. In other words, dropping a table inside a transaction and later rolling back the transaction to "undo dropping" is impossible.

For MySQL specifically, see https://dev.mysql.com/doc/refman/8.4/en/implicit-commit.html